cpp/tests/experimental/spatial/linestring_intersection_test.cu

```diff
@@ -0,0 +1,447 @@
+    std::make_unique<rmm::device_uvector<types_t>>(d_types_buffer, stream),
+    std::make_unique<rmm::device_uvector<IndexType>>(d_offset_buffer, stream),
+    std::make_unique<rmm::device_uvector<point_t>>(d_points_coords, stream),
+    std::make_unique<rmm::device_uvector<segment_t>>(d_segments_coords, stream),
+    std::make_unique<rmm::device_uvector<IndexType>>(d_lhs_linestring_ids, stream),
+    std::make_unique<rmm::device_uvector<IndexType>>(d_lhs_segment_ids, stream),
+    std::make_unique<rmm::device_uvector<IndexType>>(d_rhs_linestring_ids, stream),
+    std::make_unique<rmm::device_uvector<IndexType>>(d_rhs_segment_ids, stream)};
+}
+
+template <typename T>
+struct LinestringIntersectionTest : public ::testing::Test {
+  rmm::cuda_stream_view stream() { return rmm::cuda_stream_default; }
+  rmm::mr::device_memory_resource* mr() { return rmm::mr::get_current_device_resource(); }
+
+  template <typename IndexType, typename MultiLinestringRange, typename IntersectionResult>
+  void run_single_test(MultiLinestringRange lhs,
+                       MultiLinestringRange rhs,
+                       IntersectionResult const& expected)
+  {
+    auto got = pairwise_linestring_intersection<T, IndexType>(lhs, rhs, this->mr(), this->stream());
+
+    CUSPATIAL_EXPECT_VECTORS_EQUIVALENT(*std::move(expected.geometry_collection_offset),
+                                        *std::move(got.geometry_collection_offset));
+    CUSPATIAL_EXPECT_VECTORS_EQUIVALENT(*std::move(expected.types_buffer),
+                                        *std::move(got.types_buffer));
+    CUSPATIAL_EXPECT_VECTORS_EQUIVALENT(*std::move(expected.offset_buffer),
+                                        *std::move(got.offset_buffer));
+    CUSPATIAL_EXPECT_VECTORS_EQUIVALENT(*std::move(expected.points_coords),
+                                        *std::move(got.points_coords));
+    expect_segment_equivalent(*std::move(expected.segments_coords),
+                              *std::move(got.segments_coords));
+    CUSPATIAL_EXPECT_VECTORS_EQUIVALENT(*std::move(expected.lhs_linestring_id),
+                                        *std::move(got.lhs_linestring_id));
+    CUSPATIAL_EXPECT_VECTORS_EQUIVALENT(*std::move(expected.lhs_segment_id),
+                                        *std::move(got.lhs_segment_id));
+    CUSPATIAL_EXPECT_VECTORS_EQUIVALENT(*std::move(expected.rhs_linestring_id),
+                                        *std::move(got.rhs_linestring_id));
+    CUSPATIAL_EXPECT_VECTORS_EQUIVALENT(*std::move(expected.rhs_segment_id),
+                                        *std::move(got.rhs_segment_id));
+  }
+};
+
+// float and double are logically the same but would require seperate tests due to precision.
+using TestTypes = ::testing::Types<float, double>;
+TYPED_TEST_CASE(LinestringIntersectionTest, TestTypes);
+
+// // TODO: sort the points in the intersection result since the result order is arbitrary.
```

Comment:
Is this for this PR? Seems hard to compare `got` and `expected` if results are arbitrary?

Reply:
Yeah, this still needs to be done. Otherwise the test maybe flaky.

Reply:
Ok, I think sorting the result is finally added!